### PR TITLE
[JAVA] Prioritize mapped discriminators over generated (relates to issue #12777)

### PR DESCRIPTION
**Pull request by robbertvanwaveren**
This PR introduces a slightly smarter ordering of the discriminators which prioritizes actual non-generated mappings as defined in the original spec above others as suggested here https://github.com/OpenAPITools/openapi-generator/issues/12777#issuecomment-1299967591.

The result is that the correct value is selected to populate the discriminator field during serialization based on the actual discriminator mapping is available in the schema.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date.sh
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by robbertvanwaveren** [comment]
@wing328 Could you have a look at this change?

(the failed check is inherited from the issue in master with the python-nextgen client and is not related)

**Comment by Qunit** [comment]
Nice fix

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenDiscriminator.java [hunk 1]
```diff
@@ -128,7 +128,13 @@ public int compareTo(MappedModel other) {
             } else if (other.getMappingName() == null) {
                 return -1;
             }
-            return getMappingName().compareTo(other.getMappingName());
+
+            final boolean mappedEqualsModelName = getMappingName().equals(getModelName());
```
[the remark this answers is not in the document]
Comment (robbertvanwaveren) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenDiscriminator.java:134]: I've added the comment and rebased the branch with current master @wing328

[the remark this answers is not in the document]
Comment (robbertvanwaveren) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenDiscriminator.java:134]: rebased branch to deal with removed python samples

**Comment by b3nk4n** [comment]
Thank you! This is a really helpful fix for us!

**Comment by mcleap** [comment]
Please merge it before 7.0.1 release, we really like this feature. :) We've just faced with this issue...

**Comment by wing328** [comment]
Can someone please share a spec to reproduce the issue? I tested with the one in https://github.com/OpenAPITools/openapi-generator/issues/12777#issuecomment-1618731277 but the output remains the same with this patch/PR.

**Comment by robbertvanwaveren** [comment]
In the original issue different people actually refer to different issues. This PR solves the most generic of those. Namely that objects are serialized using the proper value as per the _key_ in the _discriminator_ spec.

I don't have time to test it right now, but I believe this spec shows the problem
```
openapi: 3.0.1
info:
  title: Discriminator Test spec
  version: 0.0.1
paths:
  /random-parent:
    get:
      operationId: test
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Parent'
components:
  schemas:
    Parent:
      type: object
      properties:
        parent:
          type: string
        child:
          oneOf:
            - $ref: '#/components/schemas/Child1'
            - $ref: '#/components/schemas/Child2'              
    Child1:
      allOf:
        - $ref: '#/components/schemas/Child'
        - type: object
          properties:
            problematic:
              type: boolean
              default: true
    Child2:
      allOf:
        - $ref: '#/components/schemas/Child'
        - type: object
          properties:
            sweet:
              type: boolean
              default: true
    Child:
      type: object
      properties:
        name:
          type: string
        type:
          type: string
      discriminator:
        propertyName: 'type'
        mapping:
          # ProblemChild/SweetChildOfMine come alphabetically behind Child1/Child2
          ProblemChild: '#/components/schemas/Child1'
          SweetChildOfMine: '#/components/schemas/Child2'
```

The resulting interface "ParentChild.java" should look like this with this patch:
```
@JsonSubTypes({
  @JsonSubTypes.Type(value = Child1.class, name = "ProblemChild"),
  @JsonSubTypes.Type(value = Child2.class, name = "SweetChildOfMine"),
  @JsonSubTypes.Type(value = Child1.class, name = "Child1"),
  @JsonSubTypes.Type(value = Child2.class, name = "Child2"),
})
```
whereas without it the ordering will be different, resulting in jackson picking the wrong values during serialization.

**Comment by wing328** [comment]
FYI @OpenAPITools/generator-core-team

**Pull request merged**